server/lib/scraper.js

```diff
@@ -0,0 +1,107 @@
+const cheerio = require('cheerio');
+const request = require('request-promise');
+const parseHtml = require('../classes/Machine').parse;
+
+/**
+ * This function scrapes a list of all of the available laundry rooms. It returns an array of objects
+ * specifying both the name and full url of each room
+ */
+async function scrapeLocations(redis) {
+	// Initialize variables
+	const url = 'http://wpvitassuds01.itap.purdue.edu/washalertweb/washalertweb.aspx';
+	const key = 'location-urls';
+
+	// Check to make sure they aren't already cached
+	let exists = await redis.exists(key);
+	if (exists === 0) {
+		// Get the page contents
+		let html = await request(url);
+		let $ = cheerio.load(html);
+
+		// Turn the `select` options into a fancy array
+		let locations = Array.from($('#locationSelector').find('option')).map(e => ({
+			'name': e.children[0].data,
+			'url': url + '?location=' + e.attribs.value
+		}));
+
+		// Cache the results. The results are cached for 1 month because they don't change frequently
+		redis.redis.set(key, JSON.stringify(locations));
+		redis.redis.expire(key, 60 * 60 * 24 * 30);
+		return locations;
+	} else {
+		// Great, they were cached! No HTTP requests today, Satan!
+		return JSON.parse(await redis.get(key));
+	}
+}
+
+/**
+ * Arrays of objects are weird to deal with, so this method takes in a location, scrapes the URLs, then uses Array.find
+ * to find our object
+ */
+async function getUrlFor(location, redis) {
+	const locations = await scrapeLocations(redis);
+	const l = locations.find(l => l.name === location);
+	return l ? l.url : '';
+}
+
+/**
+ * This method simply scrapes the locations, then goes through every one of them, scraping the machines on each page
+ */
+async function scrapeAllMachines(redis) {
```

Comment:
Why isnt this using scrapeMachinesAt() ? Is it redundant?

Reply:
huh..not sure...to be fair the whole method is getting rewritten soon since itap provides every machine at the root url anyway